YQL-18051: Add block implementation for Exists callable

igormunkin opened the pull request:
This patchset supports block implementation for `Exists` callable. All the phases implements quite similar transformations and checks to the scalar analogue. And still there is block specific optimization in kernel for the cases when all values in the batch and are invalid (i.e. `NULL`) or valid (i.e. `NOT NULL`). In this case the data buffer in the resulting array is filled by 0 or 1 respectively.

Apart from the main changeset, several helpers are introduced (or moved to the common headers) within the series.

### Changelog category

* Improvement

igormunkin committed:
YQL-18051: Add tests for block implementation

igormunkin committed:
YQL-18051: Implement type annotation phase for BlockExists

The introduced node handler implements quite the similar checks and
optimizations to the scalar analogue.

igormunkin committed:
YQL-18051: Implement computation phase for BlockExists

This patch implements the helper emitting the relevant TBlockFuncNode.
There is optimization in kernel for the cases when all values in the
batch and are invalid (i.e. NULL) or valid (i.e. NOT NULL). In this case
the data buffer in the resulting array is filled by 0 or 1 respectively.

igormunkin committed:
YQL-18051: Implement node emitting phase for BlockExists

The introduced node handler implements quite the similar checks and
optimizations to the scalar analogue.

igormunkin committed:
Introduce DecompressToSparseBitmap helper

This patch introduces the  bitmap helper to convert the particular Arrow
batch bitmap to the ui8 data set.

vitstn commented on ydb/library/yql/minikql/comp_nodes/mkql_block_exists.cpp:
```diff
@@ -0,0 +1,69 @@
+        MKQL_ENSURE(input.is_array(), "Expected array");
+        const auto& arr = *input.array();
+
+        std::shared_ptr<arrow::Buffer> bitmap = ARROW_RESULT(arrow::AllocateBitmap(arr.length, ctx->memory_pool()));
```
result is always not NULL, just pass nullptr as 0th buffer

igormunkin replied:
Rewrote this part, [introduced](https://github.com/ydb-platform/ydb/pull/3400/files#diff-d1a86e68d1bedb53992628d8e160cd0d33fd9ee0f67bed1bc5294f807a0dd178R50-R62) common helpers.

vitstn commented on ydb/library/yql/minikql/comp_nodes/mkql_block_exists.cpp:
```diff
@@ -0,0 +1,69 @@
+        return exec->Exec(ctx, batch, res);
+    });
+
+    kernel->null_handling = arrow::compute::NullHandling::COMPUTED_NO_PREALLOCATE;
```
OUTPUT_NOT_NULL

igormunkin replied:
[Fixed](https://github.com/ydb-platform/ydb/pull/3400/files#diff-f988d81370707cd0c6e1560a448e08297aacf1d87b06caf0f99346c168be1d95R42).

vitstn commented on ydb/library/yql/minikql/mkql_program_builder.cpp:
```diff
@@ -1611,6 +1611,25 @@ TRuntimeNode TProgramBuilder::BlockCoalesce(TRuntimeNode first, TRuntimeNode sec
+    auto dataItemType = dataType->GetItemType();
+
+    if (dataItemType->IsVoid()) {
+        return NewDataLiteral(false);
```
output type should be Block, this optimization has already been done for scalar Exists, please remove

igormunkin replied:
[Removed](https://github.com/ydb-platform/ydb/pull/3400/files#diff-dc2f75d5908ddf8694d8a7b07b45ffb6c4efd8d7bf93912bb3efddca5e74e168R1614-R1622).

vitstn commented on ydb/library/yql/public/udf/arrow/bit_util.h:
```diff
@@ -47,6 +47,32 @@ inline void CompressSparseImpl(ui8* dst, const ui8* srcSparse, size_t len) {
 
+inline void DecompressToSparseBitmap(ui8* dstSparse, const ui8* src, size_t srcOffset, size_t len) {
+    if (srcOffset != 0) {
+        for (ui8 i = srcOffset; i < 8 - srcOffset; i++) {
```
i < 8 ?

igormunkin replied:
Reimplemented this routine. `srcOffset` is processed [another way](https://github.com/ydb-platform/ydb/pull/3400/files#diff-b8e5dc41387f49b5eb79979cfe9dc8c2cfccf15491225b4801cbdb37fcb778d8R19-R28).

vitstn commented on ydb/library/yql/public/udf/arrow/bit_util.h:
```diff
@@ -47,6 +47,32 @@ inline void CompressSparseImpl(ui8* dst, const ui8* srcSparse, size_t len) {
+        src++;
+    }
+    while (len >= 8) {
+        ui8 slot = *src++;
```
srcOffset isn't used here and below

igormunkin replied:
Ditto.

vitstn commented on ydb/library/yql/tests/sql/suites/blocks/exists.sql:
```diff
@@ -0,0 +1,12 @@
+    never_null is NULL as is_never_null,
+    0 is NULL as zero_lit,
+    NULL is NULL as null_lit,
+FROM Input;
```
+ ORDER BY key

igormunkin replied:
[Added](https://github.com/ydb-platform/ydb/pull/3400/files#diff-9cc77145816bf7d9257825f144f68187dac94a021798e05b7ee8a44d981bb2c3R11).

igormunkin committed:
Move auxiliary Arrow Array builders to utils

Two helpers allocating Arrow Arrays with the predefined contents are
moved to the common MiniKQL header to be used later in the series.

vitstn commented on ydb/library/yql/minikql/arrow/arrow_util.h:
```diff
@@ -39,6 +40,27 @@ inline std::string_view GetStringScalarValue(const arrow::Scalar& scalar) {
     return std::string_view{reinterpret_cast<const char*>(base.value->data()), static_cast<size_t>(base.value->size())};
 }
 
+template<ui8 Value>
```
please use a normal parameter here, not a template. it doesn't affect perfomance

igormunkin replied:
[Fixed](https://github.com/ydb-platform/ydb/pull/3400/files#diff-d1a86e68d1bedb53992628d8e160cd0d33fd9ee0f67bed1bc5294f807a0dd178R43).

vitstn commented on ydb/library/yql/public/udf/arrow/bit_util.h:
```diff
@@ -15,6 +15,37 @@ inline ui8* CompressAsSparseBitmap(const ui8* src, size_t srcOffset, const ui8*
 
+inline ui8* DecompressToSparseBitmap(ui8* dstSparse, const ui8* src, size_t srcOffset, size_t count) {
+    if (srcOffset != 0) {
+        size_t offsetWords = srcOffset >> 3;
```
offsetBytes

igormunkin replied:
[Renamed](https://github.com/ydb-platform/ydb/pull/3400/files#diff-b8e5dc41387f49b5eb79979cfe9dc8c2cfccf15491225b4801cbdb37fcb778d8R20-R22).

vitstn commented on ydb/library/yql/core/type_ann/type_ann_blocks.cpp:
```diff
@@ -181,6 +181,37 @@ IGraphTransformer::TStatus BlockCompressWrapper(const TExprNode::TPtr& input, TE
+
+    // At this point BlockItem type should be either an Optional or a Pg one.
+    // All other cases should be handled in the previous transform phases.
+    if (!EnsureOptionalType(input->Head().Pos(), *blockItemType, ctx.Expr) &&
```
EnsureOptionalType will throw a visible error for the Pg type.

instead, check the blockItemType->GetKind() and emit an issue if both conditions are not met.

igormunkin replied:
[Fixed](https://github.com/ydb-platform/ydb/pull/3400/files#diff-87fab8180f9c448242603fa34704349a4537afcbf4a48b38a3ae398c07fa1c07R199-R204).

igormunkin commented on ydb/library/yql/public/udf/arrow/bit_util.h:
```diff
@@ -15,6 +15,35 @@ inline ui8* CompressAsSparseBitmap(const ui8* src, size_t srcOffset, const ui8*
+        for (ui8 i = offsetTail; i < 8; i++) {
+            *dstSparse++ = (*src >> i) & 1u;
+        }
+        src++;
```
But everything is copied in the loop above, since `offsetTail` is `0`. So the only issue is that the "offsetTail" loop can be unrolled for this bitmap byte either.

Anyway, you're right, `srcOffset != 0` should be changed to `srcOffset % 8 != 0`.

igormunkin replied:
[Updated](https://github.com/ydb-platform/ydb/pull/3400/files#diff-b8e5dc41387f49b5eb79979cfe9dc8c2cfccf15491225b4801cbdb37fcb778d8R19).

vitstn commented on ydb/library/yql/minikql/comp_nodes/ut/mkql_block_exists_ut.cpp:
```diff
@@ -0,0 +1,111 @@
+
+Y_UNIT_TEST_SUITE(TMiniKQLBlockExistsTest) {
+
+Y_UNIT_TEST(ExistsWithoutOffset) {
```
single test is enough with a loop by all possible offsets

igormunkin replied:
[Updated](https://github.com/ydb-platform/ydb/pull/3400/files#diff-f9673320d1fb57a4139ff27c7dc5ca7015f9f707599805621544a2a245199328R99-R103).
